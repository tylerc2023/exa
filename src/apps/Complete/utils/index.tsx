export * from './configs';
export * from './useKeyboardHeight';
export * from './userFactory';
