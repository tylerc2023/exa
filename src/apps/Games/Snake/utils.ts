export type BoardObject = Array<Array<0 | 1 | 2>>;

export const generateRandom = (size: number) =>
  Math.floor(Math.random() * size);

export const addFood = (
  board: BoardObject,
  used: {[key: string]: boolean} = {},
): boolean => {
  const x = generateRandom(board.length);
  const y = generateRandom(board.length);

  const combo = `${x}${y}`;
  if (Object.keys(used).length === board.length) {
    return false;
  }
  if (combo in used || board[x][y] !== 0) {
    used[combo] = true;
    addFood(board, used);
    return false;
  }
  board[y][x] = 2;
  return true;
};

export const addStarting = (board: BoardObject) => {
  const center = Math.floor(board.length / 2);
  board[center][center] = 1;
};

export const generateBoard = (size: number) => {
  const board: BoardObject = [];
  for (let i = 0; i < size; i++) {
    board[i] = [];
    for (let j = 0; j < size; j++) {
      board[i][j] = 0;
    }
  }

  addStarting(board);
  addFood(board);
  return board;
};

export const nextSnakePosition = () => undefined;
export const collision = (next: any) => next && false;
export const endGame = () => undefined;
export const eat = () => undefined;
