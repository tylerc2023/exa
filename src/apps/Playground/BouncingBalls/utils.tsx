import {Animated} from 'react-native';
import {State} from './BouncingBall';
import {CanvasDimensions} from './Canvas';

const getRandomNumber = (min: number, max: number): number =>
  Math.random() * (max - min) + min;

type GetSpeedProps = {
  state: State;
  canvas: CanvasDimensions;
  diameter: number;
  pulse: boolean;
};

export const getSpeed = ({state, canvas, diameter, pulse}: GetSpeedProps) => {
  const dMax = state.diameter >= diameter * 1.1;
  const dMin = state.diameter <= diameter * 0.9;
  // TODO: recenter so pulsing happens from middle
  const growing = dMax || dMin ? !state.growing : state.growing;
  const d = state.diameter + (!pulse ? 0 : growing ? 1 : -1);
  const xMin = canvas.x;
  const xMax = canvas.x + canvas.width;
  const yMin = canvas.y;
  const yMax = canvas.y + canvas.height;
  const xLarger = state.x + d >= xMax;
  const xSmaller = state.x <= xMin;
  const yLarger = state.y + d >= yMax;
  const ySmaller = state.y <= yMin;
  // TODO: add collision
  const xSpeed = xLarger || xSmaller ? -state.xSpeed : state.xSpeed;
  const ySpeed = yLarger || ySmaller ? -state.ySpeed : state.ySpeed;
  const x = xLarger ? xMax - d : xSmaller ? xMin : state.x + xSpeed;
  const y = yLarger ? yMax - d : ySmaller ? yMin : state.y + ySpeed;
  const dx = x + xSpeed;
  const dy = y + ySpeed;
  return {xSpeed, ySpeed, x: dx, y: dy, d, growing};
};

type GetInitialStateProps = {
  canvas: CanvasDimensions;
  diameter: number;
  speed: number;
};

export const getInitialState = ({
  canvas,
  diameter,
  speed,
}: GetInitialStateProps): State => {
  const starting = {
    x: getRandomNumber(canvas.x, canvas.x + canvas.width - diameter),
    y: getRandomNumber(canvas.y, canvas.y + canvas.height - diameter),
  };
  const growing = !!Math.floor(Math.random());
  return {
    ...starting,
    position: new Animated.ValueXY(starting),
    xSpeed: getRandomNumber(-speed, speed),
    ySpeed: getRandomNumber(-speed, speed),
    diameter,
    growing,
  };
};
