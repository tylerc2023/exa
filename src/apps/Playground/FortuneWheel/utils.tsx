type GetRandomRange = {
  min: number;
  max: number;
};

const getRandomRange = ({min, max}: GetRandomRange) => {
  return Math.floor(Math.random() * (max - min) + min);
};

type GetNewLocation = {
  minSpin: number;
  maxSpin: number;
  numOfSegments: number;
  location: number;
};

export const getNewLocation = ({
  minSpin,
  maxSpin,
  numOfSegments,
  location,
}: GetNewLocation) => {
  const spins = getRandomRange({min: minSpin, max: maxSpin}) * 360;
  const winner = Math.floor(Math.random() * numOfSegments);
  const winnerSection = Math.floor(winner * (360 / numOfSegments));
  const newLocation = location + spins + winnerSection;
  return newLocation;
};

type GetWinnerIndex = {
  location: number;
  numOfSegments: number;
  angleOfSegment: number;
};

export const getWinnerIndex = ({
  location,
  numOfSegments,
  angleOfSegment,
}: GetWinnerIndex) => {
  const deg = Math.abs(Math.round(location % 360));
  const index =
    (numOfSegments - Math.floor(deg / angleOfSegment)) % numOfSegments;
  return index;
};
