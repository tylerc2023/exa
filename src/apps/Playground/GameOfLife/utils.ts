const operations = [
  [0, 1],
  [0, -1],
  [1, -1],
  [-1, 1],
  [1, 1],
  [-1, -1],
  [1, 0],
  [-1, 0],
];

type BoardItem = 0 | 1;

export type Board = Array<Array<BoardItem>>;

export const generateBoard = (size: number, random = 0): Board => {
  const board: Board = [];
  for (let i = 0; i < size; i++) {
    board[i] = [];
    for (let j = 0; j < size; j++) board[i][j] = Math.random() > random ? 0 : 1;
  }
  return board;
};

export const determineBoardItem = (
  board: Board,
  i: number,
  j: number,
): BoardItem => {
  let neighbors = 0;
  const item = board[i][j];
  const length = board.length;
  for (let k = 0; k < operations.length; k++) {
    const [x, y] = operations[k];
    const newI = i + x;
    const newK = j + y;
    if (newI >= 0 && newI < length && newK >= 0 && newK < length)
      neighbors += board[newI][newK];
    if (neighbors > 3) break;
  }

  return neighbors < 2 || neighbors > 3
    ? 0
    : item === 0 && neighbors === 3
    ? 1
    : item;
};
