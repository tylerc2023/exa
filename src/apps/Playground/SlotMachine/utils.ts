import {Combinations, Reel, Reels} from './config';

const getWinsPerCombination = (
  combinations: Combinations,
  reels: Reels,
): Combinations => {
  const output = [...combinations];
  for (let i = 0; i < combinations.length; i++) {
    let freq = 1;
    output[i] = [...combinations[i]];
    for (let j = 0; j < reels.length; j++) {
      const symbol = output[i][j];
      const reel = reels[j];
      const sum = getSumOfReel(reel);
      freq *=
        symbol !== 'any' && typeof symbol !== 'number' ? reel[symbol] : sum;
    }
    output[i][3] = freq;
  }
  return output;
};

const getSumOfReel = (reel: Reel): number => {
  return Object.values(reel).reduce((sum, item) => sum + item, 0);
};

const getSumOfReels = (reels: Reels): number => {
  return reels.reduce((total, reel) => total * getSumOfReel(reel), 1);
};

const getReturnsPerCombination = (
  combinations: Combinations,
  winsPerCombination: Combinations,
): Combinations => {
  return winsPerCombination.reduce(
    (output: Combinations, combination, index) => {
      const payout = combinations[index][3];
      output[index][3] = combination[3] * payout;
      return output;
    },
    [...winsPerCombination],
  );
};

const getSumOfCombinations = (combinations: Combinations): number => {
  return combinations.reduce((total, combination) => total + combination[3], 0);
};

export const getWinPercentage = (
  combinations: Combinations,
  reels: Reels,
): number => {
  const sumOfReels = getSumOfReels(reels);
  const winsPerCombination = getWinsPerCombination(combinations, reels);
  const sumOfWins = getSumOfCombinations(winsPerCombination);
  return sumOfWins / sumOfReels;
};

export const getReturnPercentage = (
  combinations: Combinations,
  reels: Reels,
): number => {
  const sumOfReels = getSumOfReels(reels);
  const winsPerCombination = getWinsPerCombination(combinations, reels);
  const returnsPerCombination = getReturnsPerCombination(
    combinations,
    winsPerCombination,
  );
  const sumOfReturns = getSumOfCombinations(returnsPerCombination);
  return sumOfReturns / sumOfReels;
};
