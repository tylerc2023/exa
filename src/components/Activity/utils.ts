import axios from 'axios';
import endOfWeek from 'date-fns/endOfWeek';
import format from 'date-fns/format';
import isToday from 'date-fns/isToday';
import startOfWeek from 'date-fns/startOfWeek';
import sub from 'date-fns/sub';
import {ActivityMatrix, Site} from './types';

type ApiResponse = {[unix: string]: number};

type ApiPromise = Promise<ApiResponse>;

type ApiInput = {
  username: string;
  site: Site;
};

type ActivitySquares = {
  matrix: ActivityMatrix;
  max: number;
  total: number;
};

export const getDateFormat = (date: number): string =>
  format(date, 'yyyy-MM-dd');

export const getSubmissionFormat = (count: number, date: number): string => {
  const submissions = count === 1 ? 'submission' : 'submissions';
  const day = isToday(date) ? 'today' : `on ${format(date, 'MMM dd, yyyy')}`;
  return `${count} ${submissions} ${day}`;
};

export const getActivitySquares = (): ActivitySquares => {
  const today = Date.now();
  const matrix = [];
  const oneDay = 60 * 60 * 24 * 1000;
  const begin = sub(startOfWeek(today), {years: 1}).valueOf();
  let end = endOfWeek(today).valueOf();
  let day = 0;
  let week = [];
  while (end >= begin) {
    week.unshift({date: end, count: 0});
    if (day > 5) {
      matrix.push(week);
      week = [];
      day = 0;
    } else {
      day++;
    }
    end -= oneDay;
  }

  return {matrix, max: 0, total: 0};
};

export const updateActivitySquares = (
  squares: ActivitySquares,
  active: ApiResponse,
): ActivitySquares => {
  let max = 0;
  let total = 0;
  const matrix = squares.matrix.map((week) =>
    week.map((day) => {
      const count = active[getDateFormat(day.date)] || 0;
      max = Math.max(count, max);
      total += count;
      return {...day, count};
    }),
  );
  return {matrix, max, total};
};

const getGithubActivity = async (username: string): Promise<ApiResponse> => {
  const url = `https://github-contributions-json-api.now.sh/api?username=${username}`;
  const res = await axios.get(url);
  return res.data;
};

const getLeetCodeActivity = async (username: string) => {
  const url = `https://leetcode.com/api/user_submission_calendar/${username}`;
  const res = await axios.get(url);
  const data = JSON.parse(res.data);
  return Object.keys(data).reduce((total: ApiResponse, item) => {
    const day = getDateFormat(Number(item) * 1000);
    if (day in total) total[day] += data[item];
    else total[day] = data[item];

    return total;
  }, {});
};

const getHackerRankActivity = async (username: string) => {
  const url = `https://www.hackerrank.com/rest/hackers/${username}/submission_histories`;
  const res = await axios.get(url);
  return Object.keys(res.data).reduce((total: ApiResponse, item) => {
    total[item] = Number(res.data[item]);
    return total;
  }, {});
};

export const getApiActivity = ({username, site}: ApiInput): ApiPromise => {
  switch (site) {
    case 'github':
      return getGithubActivity(username);
    case 'leetCode':
      return getLeetCodeActivity(username);
    case 'hackerRank':
      return getHackerRankActivity(username);
    default:
      return Promise.resolve({});
  }
};
