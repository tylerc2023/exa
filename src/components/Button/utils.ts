import {StyleSheet} from 'react-native';
import {colorWithOpacity, Theme} from '../../utils';
import {Color} from './../../models/Theme';
import {ButtonEmphasis} from './types';

export const getButtonColor = (
  colorScheme: Color,
  color: keyof Color,
): string => colorScheme[color];

interface StyleInterface {
  colorScheme: Color;
  color: string;
  emphasis: ButtonEmphasis;
  noPadding?: boolean;
  disable?: boolean;
}

export const getStyles = ({
  colorScheme,
  color,
  emphasis,
  noPadding,
  disable,
}: StyleInterface): any =>
  StyleSheet.create({
    center: {
      alignSelf: 'center',
    },
    container: {
      backgroundColor:
        disable && emphasis === 'high'
          ? colorWithOpacity(color, 0.38)
          : emphasis === 'high'
          ? color
          : 'transparent',
      borderColor:
        disable && emphasis === 'medium'
          ? colorWithOpacity(color, 0.38)
          : emphasis === 'medium'
          ? colorScheme.dark
          : 'transparent',
      borderRadius: Theme.padding.p01,
      borderWidth: 1,
      flexDirection: 'row',
      padding: noPadding ? Theme.padding.p00 : Theme.padding.p02,
      paddingHorizontal: noPadding ? Theme.padding.p00 : Theme.padding.p04,
    },
    invisible: {
      opacity: 0,
    },
    label: {
      height: Theme.padding.p05,
      justifyContent: 'flex-start',
      marginVertical: Theme.padding.p01,
      paddingHorizontal: 0,
    },
    right: {
      alignSelf: 'flex-end',
    },
    text: {
      color: disable
        ? colorWithOpacity(color, 0.38)
        : emphasis === 'high'
        ? colorScheme.background
        : color,
    },
  });
