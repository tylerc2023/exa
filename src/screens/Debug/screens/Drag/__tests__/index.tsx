import * as React from "react";
import { create } from "react-test-renderer";
import { Drag } from "..";
import { MockProviderAndRouter } from "../../../../../utils";

it("renders correctly", () => {
  const dom = create(<MockProviderAndRouter Component={Drag} />).toJSON();
  expect(dom).toMatchSnapshot();
});
