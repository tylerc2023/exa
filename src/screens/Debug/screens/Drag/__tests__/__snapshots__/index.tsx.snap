// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Array [
      Object {
        "backgroundColor": "hsl(0, 100%, 100%)",
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "hsl(0, 100%, 100%)",
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "flex": 1,
          "opacity": 1,
          "padding": 12,
        }
      }
    >
      <View>
        <Text
          style={
            Array [
              Object {
                "fontSize": 14,
                "fontWeight": "300",
                "letterSpacing": 0.25,
              },
              undefined,
              undefined,
              undefined,
            ]
          }
        >
          arrow-left
        </Text>
      </View>
    </View>
  </View>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentContainerStyle={
      Array [
        Object {
          "backgroundColor": "hsl(0, 100%, 100%)",
          "flex": 1,
        },
        undefined,
      ]
    }
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="handled"
    keyboardSpace={0}
    onScroll={[Function]}
    resetKeyboardSpace={[Function]}
    scrollEnabled={false}
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <View>
      <Text
        style={
          Array [
            Object {
              "fontSize": 14,
              "fontWeight": "300",
              "letterSpacing": 0.25,
            },
            Object {
              "textAlign": "center",
            },
            undefined,
            undefined,
          ]
        }
      >
        drag the circle
      </Text>
      <View
        onMoveShouldSetResponder={[Function]}
        onMoveShouldSetResponderCapture={[Function]}
        onResponderEnd={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderReject={[Function]}
        onResponderRelease={[Function]}
        onResponderStart={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        onStartShouldSetResponderCapture={[Function]}
        style={
          Object {
            "borderColor": "black",
            "borderRadius": 30,
            "borderWidth": 30,
            "height": 60,
            "left": 375,
            "marginLeft": -30,
            "marginTop": -30,
            "top": 667,
            "width": 60,
          }
        }
      />
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
